oeps/oep-0026-arch-xapi-integration.rst

```diff
@@ -0,0 +1,421 @@
+~~~~~~~~~~~
+
+The **Result** field specifies the score the user earned on an activity.  Here is an
+example of a JSON value that we would generate for a problem type:
+
+::
+
+    "result": {
+        "success": false,
+        "completion": true,
+        "score": {
+            "min": 0,
+            "max": 50,
+            "raw": 10,
+            "scaled": .20
+        }
+        "response": "foo"
+    }
+
+See `Deep Dive: Result`_ for more information on xAPI Result.
+
+.. _`Deep Dive: Result`: https://xapi.com/deep-dive-result/
+
+Open edX xAPI Plugin
+====================
+
+A new library will be implemented, using the Open edX `Django App Plugin`_ design pattern,
+that:
+
+1. Translates Open edX native events to Open edX xAPI events (**Translator**).
+2. Validates the generated event using a general-purpose validation library 
+   cross-maintained by Open edX xAPI consumers (**Validator**).
+3. Routes real-time xAPI events to all registered consumers (**Router**).
+4. Provides an administration interface to configure and manage registered
+   consumers (**Admin**).
+
+Translator
+~~~~~~~~~~
+
+This layer will plug into the `Open edX Event Tracking`_ subsystem to listen to
+events of interest and translate them to xAPI's JSON format.
+
+TBD - The development team will assess whether we should use (and start owning)
+the already implemented (but no longer maintained) `xAPI Python Open Source Library`_. 
+
+Validator
+~~~~~~~~~
+
+The intention of this validation layer is to ensure we continue to support the
+expected xAPI formats by all participating Open edX xAPI consumers. In many ways,
+this is similar to what `consumer-driven contract testing`_ would enable and uses
+similar design principles. Essentially, a common validation library can be
+collaboratively maintained by participating consumers, including consuming adaptive
+engines. This ensures an end-to-end integration that can be maintained going forward.
+
+**Note**: Although xAPI specifies a standardized format, it is a low-level transaction
+schema and relies on higher-level "profiles" applied on top of it. So the profiles
+for specific Activities, Verbs, Contexts, etc used by Open edX need to be contractually
+supported.
+ 
+Router
+~~~~~~
+
+The router layer forwards xAPI events to interested consumers.  The complexity of this
+layer may increase over time as we support higher degrees of customization by consumers,
+such as filtering for only certain types of events and varying degrees of permissions
+of access to activities.
+
+For the first iteration, we need the following permissions:
+
+* Course restriction - certain consumers can access events only in certain courses.
+
+In the future, we may need the following:
+
+* User restriction - certain consumers can access all events for certain users.
+* Site restriction - certain consumers are limited to accessing events of certain sites.
+* Activity type restriction - certain consumers can access only certain types of events.
```

Comment:
Do you think it must be a layer of the app? IMO it must be a separated component acting as a proxy to forward the right event to the right consumer with appropriate permissions.

Reply:
To be clear, when I use the term "layer" here, I simply mean a separable logical part of the plugin.  I do not mean to imply a specific way of how it is implemented.  

Would it be clearer if I use the term "component" instead of "layer"?